types/wlr_keyboard_shortcuts_inhibit_v1.c

```diff
@@ -0,0 +1,212 @@
+#include <assert.h>
+#include <stdlib.h>
+#include <util/signal.h>
+#include <wlr/types/wlr_keyboard_shortcuts_inhibit_v1.h>
+#include "keyboard-shortcuts-inhibit-unstable-v1-protocol.h"
+
+static const struct zwp_keyboard_shortcuts_inhibit_manager_v1_interface
+	keyboard_shortcuts_inhibit_impl;
+
+static const struct zwp_keyboard_shortcuts_inhibitor_v1_interface
+	keyboard_shortcuts_inhibitor_impl;
+
+static struct wlr_keyboard_shortcuts_inhibit_manager_v1 *
+wlr_keyboard_shortcuts_inhibit_manager_v1_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwp_keyboard_shortcuts_inhibit_manager_v1_interface,
+		&keyboard_shortcuts_inhibit_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static struct wlr_keyboard_shortcuts_inhibitor_v1 *
+wlr_keyboard_shortcuts_inhibitor_v1_from_resource(
+		struct wl_resource *resource) {
+	assert(wl_resource_instance_of(resource,
+		&zwp_keyboard_shortcuts_inhibitor_v1_interface,
+		&keyboard_shortcuts_inhibitor_impl));
+	return wl_resource_get_user_data(resource);
+}
+
+static void keyboard_shortcuts_inhibitor_v1_destroy(
+		struct wlr_keyboard_shortcuts_inhibitor_v1 *inhibitor) {
+	if (!inhibitor) {
+		return;
+	}
+
+	wlr_signal_emit_safe(&inhibitor->events.destroy, inhibitor);
+
+	wl_resource_set_user_data(inhibitor->resource, NULL);
```

Comment:
We should probably implicitly send a deactivate event here.

Reply:
Mhmm, isn't destroying an inhibitor in its strictest sense something different than deactivating it? My understanding is that a client creates or destroys an inhibitor while the server is only able to (de-)activate but *not* destroy it. Based on that, wouldn't the client obviously assume the inhibitor to become irrelevant/inactive after being destroyed? Couldn't it even be confusing to a client state machine to receive an inactive event while it's destroying it?

As prior art, pointer contraint doesn't send unconfined/unlocked events on destroy either.

The protocol has no wording on it either way.

Reply:
I'm worried about the case where the seat or surface gets destroyed, and this implicitly makes the inhibitor inert. In this case the client won't know that the inhibitor is no longer active, yet the inhibitor object will still be alive.

Reply:
Ah, then we could send the notification from the respective resource destroy handlers only: https://github.com/swaywm/wlroots/pull/2026/commits/44619800af73169f2bae283f8e252cd13edb3352.

Should pointer constraints do this as well then?

Reply:
Probably yeah!

Reply:
I had a look but it's not as simple since constraints can be updated (set_region). So tracking if it's active and a notification should be sent is not as straightforward. Should I even attempt to muck about in there? :)

Reply:
No worries, I've created an issue for this: https://github.com/swaywm/wlroots/issues/2038